### PR TITLE
Eliminate firrtl.node ops when they carry no annotations

### PR DESCRIPTION
**Pull request by darthscsi**
Eliminates unused NodeOps.
Partially addresses #299

### CODE REVIEW

**Comment by darthscsi** [comment]
@seldridge to make sure this doesn't break annotations.

**Comment by darthscsi** [comment]
#900 needs this to fold away the muxes we are seeing in practice.

**Review by lattner — APPROVED** [review]
awesome

File: test/Dialect/FIRRTL/canonicalization.mlir [hunk 1]
```diff
@@ -1274,4 +1274,11 @@ firrtl.module @MuxCanon(%c1: !firrtl.uint<1>, %c2: !firrtl.uint<1>, %d1: !firrtl
   // CHECK: firrtl.mux(%c1, %d2, %d1) : (!firrtl.uint<1>, !firrtl.uint<5>, !firrtl.uint<5>) -> !firrtl.uint<5> 
 }
 
+// CHECK-LABEL: firrtl.module @EmptyNode
+firrtl.module @EmptyNode(%d1: !firrtl.uint<5>, %foo: !firrtl.flip<uint<5>>) {
+  %bar = firrtl.node %d1  : !firrtl.uint<5>
+  firrtl.connect %foo, %bar : !firrtl.flip<uint<5>>, !firrtl.uint<5>
+}
+// CHECK: firrtl.connect %foo, %d1
+
```
Comment (seldridge) [inline test/Dialect/FIRRTL/canonicalization.mlir:1283]: Can you add a test of the non-empty attribute dictionary not being eliminated?

**Review by seldridge — APPROVED** [review]
lgtm

I didn't test this, but the attribute blocking approach to prevent annotations from getting removed makes sense to me.

File: include/circt/Dialect/FIRRTL/Canonicalization.td [hunk 2]
```diff
@@ -19,6 +19,11 @@ include "circt/Dialect/FIRRTL/FIRRTL.td"
 /// const-on-LHS rewriting patterns converge in case both operands are constant.
 def NonConstantOp : Constraint<CPred<"!$0.getDefiningOp<ConstantOp>()">>;
 
+// Constraint that matches empty attribute dictionaries.  Used to ensure that
+// there are no FIRRTL annotation on an operation.
+def EmptyAttrDict : Constraint<CPred<"$0.empty()">>;
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/Canonicalization.td:24]: Also, this is an _extremely_ clever solution!

**Pull request merged**